packages/realm/src/app-services/SyncSession.ts

```diff
@@ -229,106 +228,90 @@ const mockApp = new Proxy({} as App, {
 >({
   throwOnReAdd: true,
-  add(callback, internal, direction, mode) {
+  add(callback, weakInternal, internal, direction, mode) {
     const token = internal.registerProgressNotifier(
       (transferred, transferable) => callback(Number(transferred), Number(transferable)),
       toBindingDirection(direction),
       mode === ProgressMode.ReportIndefinitely,
     );
-    return { internal, token };
+    return { weakInternal, token };
   },
-  remove({ internal, token }) {
-    return internal.unregisterProgressNotifier(token);
+  remove({ weakInternal, token }) {
+    weakInternal.withDeref((internal) => internal?.unregisterProgressNotifier(token));
   },
 });
 
 /**
  * Connection listeners are shared across instances of the SyncSession, making it possible to deregister a listener on another session
  * TODO: Consider adding a check to verify that the callback is removed from the correct SyncSession (although that would break the API)
  */
-const CONNECTION_LISTENERS = new Listeners<ConnectionNotificationCallback, ListenerToken, [binding.SyncSession]>({
+const CONNECTION_LISTENERS = new Listeners<
+  ConnectionNotificationCallback,
+  ListenerToken,
+  [binding.WeakSyncSession, binding.SyncSession]
+>({
   throwOnReAdd: true,
-  add(callback, internal) {
+  add(callback, weakInternal, internal) {
     const token = internal.registerConnectionChangeCallback((oldState, newState) =>
       callback(fromBindingConnectionState(newState), fromBindingConnectionState(oldState)),
     );
-    return { internal, token };
+    return { weakInternal, token };
   },
-  remove({ internal, token }) {
-    internal.unregisterConnectionChangeCallback(token);
+  remove({ weakInternal, token }) {
+    weakInternal.withDeref((internal) => internal?.unregisterConnectionChangeCallback(token));
   },
 });
 
 export class SyncSession {
-  private static instances = new Set<binding.WeakRef<SyncSession>>();
-
-  /**
-   * Resets the internal shared pointer of all instances returned since this message was last called.
-   * @internal
-   */
-  public static resetAllInternals() {
-    assert(flags.ALLOW_CLEAR_TEST_STATE, "Set the flags.ALLOW_CLEAR_TEST_STATE = true before calling this.");
-    for (const sessionRef of SyncSession.instances) {
-      sessionRef.deref()?.resetInternal();
-    }
-    SyncSession.instances.clear();
-  }
-
   /** @internal */
-  private _internal: binding.SyncSession | null;
+  private weakInternal: binding.WeakSyncSession;
   /** @internal */
-  public get internal() {
-    assert(this._internal, "This SyncSession is no longer valid");
-    return this._internal;
+  public withInternal<Ret = void>(cb: (syncSession: binding.SyncSession) => Ret) {
+    return this.weakInternal.withDeref((syncSession) => {
+      assert(syncSession, "This SyncSession is no longer valid");
+      return cb(syncSession);
+    });
   }
 
   /** @internal */
   constructor(internal: binding.SyncSession) {
```

Comment:
I wonder if it's a problem that a strong `binding.SyncSession` is ever returned to the SDK?

Reply:
I _think_ everywhere we touch a `binding.SyncSession` it basically immediately goes through here. In particular, I was concerned about these, and considered adding a call to `weaken()` immediately after getting a `binding.SyncSession` out of native, but I realized that it would still happen at exactly the same point, so didn't seem worth it. https://github.com/realm/realm-js/blob/fb7de6e1367a48a91f1892cfa90885537450c346/packages/realm/src/app-services/Sync.ts#L74-L87

Reply:
After writing this comment, I understood that `weaken` will call the `$resetSharedPtr` which makes me less worried about a potential issue.